### ISSUE TITLE
Can you add one more demo conversation plz,

### ISSUE THREAD

**Issue by OlegKostin** [wizered67/ProjectVisualNovel#3]
I want to understand how it will wokr one with another one. And plz give me some tips about conversations .conv, all that tags.Are you writing *.conv files  manually? How can i use ConversationSchema.xsd? Thanks.

**Comment by wizered67** [comment]
I can add another example within the next few days. Ideally I'd also like to add some more thorough documentation soon. In the mean time, consider looking at the somewhat decent documentation in XML Documentation.xml. 

I am writing the .conv files by hand, but the schema adds autocompletion for most tags which is very helpful. If you are using IntelliJ, then I believe you can get the schema working Settings->Languages and Frameworks->Schemas and DTDs, clicking the green plus button, and selecting the .xsd file provided.

Is there anything in particular you would like more help/documentation for? Most of it seems pretty intuitive for me but no doubt that's because I know how everything works. Do you understand branches?  Probably the most complicated thing is using scripting languages, so I can definitely provide more support with that.

**Comment by OlegKostin** [comment]
I'm using Android Studio but its practically the same and your recommendation is working. Thanks!
I think i can figure it out by myself.   Im appreciate with all your help and hope i can ask u some point which  will be hard for me in future.
 Can you add command to add pics on backscreen? Or maybe do u have it already? Сouldn't find it out.
Thanks for your working.

**Comment by OlegKostin** [comment]
where r u generate bottom area with text? I need to add font style there in order to add russian letters.
For example: Label label = new Label("", new Label.LabelStyle(font, new Color(1, 1, 1, 1))); 
Is it possible?

**Comment by wizered67** [comment]
The labels for text are all created in the constructor of GUI Manager. Take a look at that and you should be able to find what to modify.

I haven't gotten around yet to adding a command for background images but I'll put it on my list and try to get to it soon, possibly tonight or the next few days. Also, I'm just curious, what made you want to use this for your game instead of more developed/well known alternatives?

**Comment by wizered67** [comment]
I have added a command to set a background image. There's one example in the updated demonstration but essentially you just use <bg image="filename of image, including directory and file extension">. To add your own backgrounds you will need to make sure they get loaded. To do this, check out the loadDefault method in GameManager.java. You'll just need to add another call to assetManager.load with the resource you add.

A few notes:
- The asset loading system is kind of a mess right now. At some point (hopefully soon) I'd like to streamline it and make it easier.
- I haven't correctly set up viewports yet so you'll notice that resizing the game window scales the screen, which can distort it. I've put off fixing this until the future, but at some point I'll correct this. If you want to change the default screen size you can change DEFAULT_WIDTH and DEFAULT_HEIGHT in Constants.java.
- I'm still learning how to create XSD schemas, so I hadn't notice that I accidentally left the namespace as a test name (test.Namespace. Oops). I corrected that in this version so that the namespace is "conversations". I have therefore updated the schema as well as the demonstration. Make sure that when you make .conv files you include xmlns="urn:conversations" in the conversation tag.

Let me know if you have any more questions, have found bugs, or want to request a feature!

**Comment by OlegKostin** [comment]
Thanks. 
I choosed your project because im learing libgdx and android by myself. No point in taking well known alternatives for me. And thanks again for your working.)

**Comment by OlegKostin** [comment]
Ive added russian letters, ty. 
I havent understood some parts of xml logic. So i have some questions.
How can i use variable operations in that xml form like flag. To turn variable from game logic to false or true or to increment it i++?
Can u add one more .conv file? I want to understand how can i get sequence of convisations.
How r u rendering pics on background?  It is taking left bottom part of my pic. As u said u r gonna fix it, but maybe i can help u a little.

**Comment by wizered67** [comment]
For the background being the left bottom, I believe that is related to the origin being in the bottom left corner for textures. I'm a bit confused about why it would be doing it in a problematic way. Can you attach the image you're trying to use so I can take a closer look? 

As for the sequence of conversations question, I assume you're referring to going from one conversation to another? I haven't gotten around to making a good way to do that yet, so there's no command. You should be able to do it with code with something like `GUIManager.conversationController().loadConversation("demonstration.conv");
 GUIManager.conversationController().setBranch("default");`

Using the scripting languages is a little complex, so I probably need to make a more thorough guide. You can't change Java variables directly in conversations, but you can introduce variables specific to the scripting language. For example, in the demonstration conversation I use an <init> command to create a new scripting variable called "work" and set it to 0 if it hasn't been initialized yet. Later on I can use a <script> command to increment it with "work += 1".  The "work" variable can be referenced throughout the conversation, or even other conversations since all scripting variables are global variables. If you need to access a scripting variable in Java code, you can use `GUIManager.conversationController().scriptManager("language name here").getValue("variable name here")` 
This still needs to be converted to a Java type however, which can be done with a ScriptManagers objectToString and objectToBoolean methods. It's an oversight that there's no objectToInteger method so I can add that later. Anyway, writing this out made me realize how complicated it is so I will see if I can simplify it!

Also, are you planning on developing for Android with this? At the moment, Lua is the only scripting language that supports Android. I definitely need to write a more thorough guide though.

**Comment by OlegKostin** [comment]
`For the background being the left bottom, I believe that is related to the origin being in the bottom left corner for textures. I'm a bit confused about why it would be doing it in a problematic way. Can you attach the image you're trying to use so I can take a closer look?`
This is cause of 	public final static int DEFAULT_WIDTH = 1024; //1024
	public final static int DEFAULT_HEIGHT = 768; //576 If u take picture with bigger picture resolution, it will trim.
Ive already had this problem in my test platformer, and fixed. I think i can fix it here either.

I dont know script languages so im starting to learn them. And by the way we cant launch yourr app on android cause u r using groovy, As u said we need Lua to launch it on android
Thanks for detailed answer, ill try to  figure it out.

**Comment by wizered67** [comment]
Just wanted to let you know that I've been working on overhauling the asset loading/managing system. Ideally this should make it much easier to add new resources AND will add more autocompletion to writing conversations. I am, however, changing the names of some attributes in some of the tags. I'll give more specific information when I'm done with this, but I just wanted to give you some heads up that some things will be changing somewhat drastically.

**Comment by OlegKostin** [comment]
Hello again. Will your schema branch release any soon? Will it support android? Thanks for your work!

**Comment by wizered67** [comment]
Hi! I've been hard at work for the last month or so adding a bunch of new features! I originally intended the schema branch to focus only on overhauling the asset management system, but I ended up adding much much more. I think I am almost ready to merge the schema branch into master, after a little more testing is done. However, I'm hesitant to call anything an official "release". A lot of things are still in development and are subject to change as improvements are made. You're welcome to use this, but just keep in mind that a lot of things may change. 

Since you mentioned it before, I have made sure to support android as much as possible. Previously I had intended Groovy to be the primary scripting language, but because Lua works on Android I intend to prioritize it instead. All of the changes I've made in the last month do seem to work on Android, as long as Lua is used. However, I have not yet implemented viewports for resolution independence. As a result, the screen will really only display properly at the moment on desktop and at the default size. This is something I intend to address in the future, but I'm still learning the best ways to approach it so it may take a while. 

Once I merge the schema branch into master, hopefully soon, I will give a more in depth explanation of all the changes that have been made. I still intend to write a better guide to making conversations at some point, but it isn't something I've gotten around to yet. Thanks for your continued interest!

**Comment by OlegKostin** [comment]
Ok,  Im appreciative for your work and waitng your new release.

**Comment by wizered67** [comment]
I have merged the Schema branch into the master branch! This is a really large and significant update that makes a lot of important changes and really puts the project on a nice track going forward. I'll do my best to go over the most important changes.

First and foremost, I want to mention that I've added documentation to the Demonstration.conv file! There are now comments explaining what a lot of the lines do. This should make it much easier to understand and help you make your own.

## Resource Management

The first major focus of this update was overhauling the resource system. Previously, all resource loading had to be hardcoded. This really made it a mess to add new resources. Additionally, using resources in the .Conv files was a pain because resource names did not autocomplete. I believe the new system fixes all of these problems. In the Assets folder, there is a new file called Resources.xml. This file allows you to define which assets will be used in the game. There are sections for different types of resources. Most allow you to add define your own assets by adding a line in the section of the form `identifier "filename"`. For example, let's say you want to add a custom background. You have a file "forest.png" which has been placed in the Textures directory. To include that asset, you would need to open Resources.xml and add a line in the `<textures>` section. That line might look something like `forestBackground "forest.png"`. (Note: the file path is relative to the Textures directory because this line is being added to the `<textures>` section, so you don't need to write it as "Textures/forest.png"). The part before the filename, 'forestBackground', is an optional identifier. Identifiers are intended to give simpler or easier to understand names to assets. If an identifier is not given, the filename is treated as its own identifier. 

It is technically not necessary, but is highly recommended to run the provided jar file,  ResourceSchemaGenerator.jar every time you make changes to Resources.xml. Note: Make sure to set the jar file's working directory to the assets folder or it won't find Resources.xml! What this jar does is automatically generate a schema file based off of the resources listed in Resources.xml. The result of this is that you can use autocompletion for resource names when writing conversations! This can significantly increase the speed of writing conversations and decrease the rate of error! Other types of resources are specified in Resources.xml as well, including characters and animations! 

Adding to Resources.xml is not enough to load the assets, however. Instead, load groups must be created in Resources.xml. See the "common" group as an example. By default, the common group is loaded when the game starts, but other groups can be loaded manually through code or through conversations. The code to do it is `GameManager.assetManager().loadGroup("name")` and the conversation command is `<loadGroup>name</loadGroup>`. When a group is loaded, all assets in that group are loaded too. Groups can be unloaded similarly, with `GameManager.assetManager().unloadGroup("name")` and `<unloadGroup>name</unloadGroup>`.

## Images!
Images can now be displayed in conversations! Images are non animated textures that can be shown at different positions. They can be moved to different positions and fade in and out. Some basic examples are in the demonstration.

## Tweening
It is now possible to "tween" (interpolate) between character's and image's positions over a set period of time. There are also many different types of interpolation that can be used for different effects. These different types are also available for fading, although they're less useful.

## Debug Menu
Pressing "Enter" while Constants.DEBUG is true will open a debug menu that lets you open conversations, change branch, save, and load. These can be helpful when debugging!

## Transcript
Pressing the right "Shift" key opens a scrollable transcript displaying a subset of the last messages that were displayed, This is helpful if the player accidentally skipped over something and needs to check it again!

## Better scripting support!
It is now easier to get and set the values of Lua variables from Java! To set a value, use `ConversationController.scriptManager("Lua").setValue(name, value)` where name is the variable name and value is a Java value. Similarly, you can get a value with `ConversationController.scriptManager("Lua").getValue(name, type)` where type is the type of the variable in Java types. For example, to get an integer it would be Integer.class.

After looking into it, it should be possible to access Java classes in Lua! For example, `system = luajava.bindClass('java.lang.System')` will set the Lua variable system as a reference to java.lang.System. From there, you can access its methods in Lua with something like `system:currentTimeMillis()`. For convenience, I have added a class in the scripting package called LuaGameMethods. You can add your own static methods to this class and then call them in Lua with `game:methodName`. Feel free to experiment with this, there are a lot of possibilities!

## Misc New Conversation Commands
(Need more documentation later)
`<fade>` can be used for a whole screen fade/flash.
`<waitUntil>` can be used to wait for a script's condition to be met before advancing.
`<waitForLoading>` can be used to wait for all queued assets to be loaded before advancing.
`<else>` can be put inside an `<if> `tag to execute commands if the original condition was not met.
`<exit> `stops execution of the current conversation and hides all GUI elements.
`<clearScene>` can remove all characters and or images being displayed




That probably sums up the most important changes, although I may have forgotten some things. I want to write more extensive documentation eventually, but for now look at the documentation added to demonstration.conv for examples of what it does! Let me know if you have any questions or comments!

**Comment by OlegKostin** [comment]
good job!!

**Comment by OlegKostin** [comment]
Hello. Here is one problem that i cant resolve.
Background image is cut off by  DEFAULT_WIDTH and DEFAULT_HEIGHT, so if it has dimension more than 800-600 it will be cut off. We need to use FitViewport to avoid it. But i can not find the Initialization of this recourse. 
Help me plz.

**Comment by wizered67** [comment]
Hi,

I assume you're using the version from the master branch? I believe that one does not have any viewports set up, which is probably the cause of the problem. I've been busy the past month so I hadn't done much work, but I had been working on adding viewports and other changes in the resolutions branch. I should have some time for now so hopefully I can finish some of the changes and push it to the master branch. Hopefully that will fix your problem! 

Thanks for your patience! If there are any other features you need, let me know!

**Comment by OlegKostin** [comment]
Thanks for your help.
Yes, i have one more problem. How can i integrate mini game like flappy bird in conversation, Here will be boolean variable. If the player get success in game, the variable get equal to true and go to one branch and with false goes to another.

**Comment by wizered67** [comment]
Good question! That should definitely be possible with the features I've set up in the last version. Here's the rough idea of what you'd want to do. 

First, open the LuaGameMethods class in the scripting package and add a static method to it, something like startMinigame. The LuaGameMethods class is just a set of methods you can define (with some random test examples already in it) and then easily call in conversations by using game:methodname() in a script command. 

Then, add something like this to your conversation. 
`<script>minigameResult = nil</script>`
`<script>game:startMinigame()</script>`
` <waitUntil>minigameResult ~= nil</waitUntil>`
`<Add your if command here and changebranches as appropriate>`
Let me explain how this works. First, a variable called minigameResult is set to nil. Next, the startMinigame method in the LuaGameMethods class is called, which will actually start the minigame. The line after that, waitUntil, ensures that the execution of the commands after it stops until the condition is met, in this case it waits until minigameResult is not nil.

All that's left is to set the variable to true or false at the end of the minigame, which you can do with the following line of Java code.
`ConversationController.scriptManager("Lua").setValue("minigameResult", true/false);`

**Comment by wizered67** [comment]
I just merged the resolutions branch into master and pushed the changes! By default it uses an ExtendViewport which seems to work pretty well. In the Constants class, there are variables for virtual width and virtual height. Those variables affect how much of the game world is shown. There are a couple of bugs which I will address later - resizing while text is in progress can cause the text to have issues which resolve when the next text is displayed. Additionally, saving doesn't fully work because text progress isn't saved. Again, I'll make sure to deal with these at some time later. Also, I don't think I mentioned it but the last version added some cool new features! I'm now using Rafaskb's wonderful typing label, which fixes some of the bugs with my implementation and adds some cool new features! Check out the available tokens! https://github.com/rafaskb/typing-label/wiki/Tokens 
I also added cameras, which can be moved and zoomed with commands! See the demonstration file for a couple examples! 

Thanks for your patience and let me know if you have any questions or suggestions!

**Comment by OlegKostin** [comment]
Thanks for explanations. Im appreciate that u also making small fixes, like method initFonts(), really nice for non english speakers, since Libgdx doesnt support another letters than english.
Viewport is working by editing VIRTUAL_WIDTH and VIRTUAL_HEIGHT now, thanks.
Now im figuring out how camera is working)

**Comment by OlegKostin** [comment]
Hmm ok ive got some silly questions..
Is it necessary to put all my resources to ResourcesSchema.xsd?
How can i start from my own convesation, im typing in MainGame.java
```
setScreen(new LoadingScreen(new LoadingScreen.LoadResult() {
			@Override
			public void finishLoading() {
				GameManager.game.setScreen(gameScreen);
				conversationController.setConv(GameManager.assetManager().getConversation("demonstration.conv"));
				conversationController.setBranch("default");
			}
		}));
```
but it says  "Asset not loaded" how can i load all my resources? Cause its impossbile to add one more background or any other res either.

What for does it need in MainGame.java 
`GameManager.assetManager().loadGroup("common");`
As i remember if ill get many backgrouunds pics i need to put them to TextureAtlas to avoid memory leak. Is it possible here?

Can u add BufferedReader commands to get  players name and use it later as one of character?

**Comment by wizered67** [comment]
Let me try to answer all your questions.
> Is it necessary to put all my resources to ResourcesSchema.xsd?

I don't think there's currently a file called that, are you referring to ConversationResources.xsd? Either way, the idea with resources is that you define them in Resources.xml. Then, running the provided ResourceSchemaGenerator.jar will automatically modify ConversationResources.xsd allowing you to use autocompletion when dealing with resources in .conv files! I know the way resources are handled at the moment is somewhat confusing and I definitely want to make it clearer by adding documentation or changing how some of it works.

> how can i load all my resources?

Let me try to explain how loading resources works. Instead of loading or unloading one resource at a time, resources are bundled into "Groups". By default, the game loads all resources in the "common" group, which is the purpose of the line `GameManager.assetManager().loadGroup("common")`. At the bottom of Resources.xml is a section with the <groups> tag. Inside of that you can add resources to the common group or add new groups of your own. Again, anything in the common group will be loaded when the game starts, but if you make your own group you will need to manually load it either with code like the line above or with the <loadGroup> command. You mentioned having trouble with your conversation not being loaded. To fix that, go to the bottom of Resources.xml and add inside the common group `<load>YourConversationName.conv</load>`.

> As i remember if ill get many backgrouunds pics i need to put them to TextureAtlas to avoid memory leak. Is it possible here?

Currently only animations are loaded from TextureAtlases, but I will put this on my to do list!

> Can u add BufferedReader commands to get players name and use it later as one of character?

I'm not entirely sure on what you mean by using BufferedReader. I assume what you're asking is if it's possible to set a character's name to some value entered by the player? Currently there's no way to do it with commands, although I'm working on a way of using Lua variables as command attributes which would allow this! You could probably do it with code like `SceneManager.characterDefinitions().get("character identifier in lowercase").setKnownName(playerName)`. I should mention that there is a bug in the version you have where the command to set a character's display name is not working! It's an easy fix and I'll definitely have it resolved once I push the next version.

I hope that cleared some things up and let me know if not!

**Comment by OlegKostin** [comment]
Thanks but i have some more questions to make it really clear.
There is no tag like convesation there. But u r  adding converation is groups

```
<groups>
        <group name="common">
  <load>demonstration.conv</load>
        </group>
    </groups>
```
What all other tags then like  <animations> or    <sounds> for? We r loading "testbackground.png" 2 times in textures and groups, why do we need it?What will happen if ill create new directory "backgrounds" and add all bg pics there? Is it neccasary to creater new tag and how?

> I'm not entirely sure on what you mean by using BufferedReader. 

U assumed right ,my mistake. That approach you suggested  is hard for me so ill just do it later.)

**Comment by OlegKostin** [comment]
and one more question plz.
Will previous background, music or sound dispose after it had been changed to next one?
Cause i didnt find that at commands and dont really know how to make it.

**Comment by wizered67** [comment]
The other tags are for listing the resources and optionally giving it a special identifier. The reason they are necessary is so that the AssetManager knows what type the resources are. For example, it knows everything in the <textures> tag should be loaded as a texture, everything in the <music> tag as music, etc. The reason it isn't necessary for conversations is just a quirk of how I programmed it (they're loaded through a different method specifically for conversations). I see how this is confusing and there's a chance that at some point I'll remove this method of doing things.

> We r loading "testbackground.png" 2 times in textures and groups, why do we need it?

So like I was saying, you can think of the first time in textures as a definition of the resource. Putting it in the groups section adds it to a group to be loaded. Resources can actually be in multiple groups. The idea is that groups are sets of resources that are frequently loaded together. For example, if at one point you go to a city you would want to load all resources associated with the city scene. You would make a city group and add to it things like the relevant conversations, the city background, the city sound effects, etc. Then you can load all of them at once with <loadGroup>. That is the purpose of the groups.

> What will happen if ill create new directory "backgrounds" and add all bg pics there?

Right now the directory format is pretty rigid, meaning you won't be able to create a new directory for backgrounds. Thanks to your feedback I'm definitely getting a sense of some of the weaknesses of the resource system, so I may change it in a future version to allow things like this.

> Will previous background, music or sound dispose after it had been changed to next one?

What do you mean by dispose? If you mean disposed from memory, then no, the loaded resources are not freed from memory until you use <unloadGroup> to do so. If you're asking whether setting a background replaces the old one, then yes. Setting backgrounds and music will replace the old one entirely. This is not the case for sounds, which at the moment cannot be stopped. The reason being that sounds are intended for short sound effects.

**Comment by OlegKostin** [comment]
Ok thanks for answers, but still one more question how can i add .conv file to Resources.xml? There is no such tag and cant add to groups either cause of error. U said that no need to put it at ConversationResources.xsd so cant resolve the problem(

> What do you mean by dispose? 
As i know in Libgdx u r need to free recourses by calling dispose() to free memory. Will i get huge memory leak if ill get for example 100 different background?

**Comment by wizered67** [comment]
What error are you getting when adding it to the group? You should just need to add it to the group and it will work fine. 

> As i know in Libgdx u r need to free recourses by calling dispose() to free memory.

The project uses AssetManager, which should dispose resources once they are unloaded. So you would need to use <unloadGroup> once you're done using the resources in it to free the memory.

**Comment by OlegKostin** [comment]
Oh my bad, ive made my conversation wrong. Ive overwrite yours democonv on mine and everything seems fine.
And ive forgot about AssetManager, ok ty for your support. Im really appreciate for that

**Comment by wizered67** [comment]
Sorry, I just noticed a typo in my previous message. I meant to say that to unload from the AssetManager you'll need to use the `<unloadGroup>` command. Just how loadGroup loads all the resources in a group, unloadGroup unloads all the resources in a group.

**Comment by OlegKostin** [comment]
Oh ok so it isnt auto, thanks. Then one more question. How many groups will i need if for example i have 100 background pics? As i get i need to sort them by chapters and use  <unloadGroup> commands after chapters has passed. Right? So i need to load all of them at start? Can i load some of them in conversations somehow?

**Comment by wizered67** [comment]
Right, so split them into groups by chapters/scenes/however you want. But you don't want to load all the groups at the beginning! Instead you load them in during conversations when you need them. For example, let's say you have a group for each chapter. At the end of chapter 4 you switch to chapter 5, so whenever you're ready to do that you can do something like `<unloadGroup>chapter4</unloadGroup>` and then `<loadGroup>chapter5</loadGroup>`. This will unload all the resources in the chapter 4 group and load all the resources in the chapter 5 group. The idea is that you only load groups when you need them and then unload them when you're done.

**Comment by OlegKostin** [comment]
Now i see, thanks for your patience!

**Comment by OlegKostin** [comment]
Where should i start from. Ive got error.
Exception in thread "LWJGL Application" com.badlogic.gdx.utils.GdxRuntimeException: Asset not loaded: Conversations/intro.conv
This is MainGame.java
```
GameManager.assetManager().loadGroup("intro");
		gameScreen = new MainGameScreen();
		setScreen(new LoadingScreen(new LoadingScreen.LoadResult() {
			@Override
			public void finishLoading() {
				GameManager.game.setScreen(gameScreen);
				conversationController.setConv(GameManager.assetManager().getConversation("intro.conv"));
				conversationController.setBranch("intro");
```
intro.conv
```
<conversation xmlns="urn:conversations">
    <branch name="intro">
        <bg id="intro0.0.png"/>

    </branch>
</conversation>
```

Resources.xml
```
    <textures>
        icon "testicon.png"
        "testbackground.png"
        icon2 "testicon2.png"
        "intro0.0.png"
    </textures>
<groups>
        <group name="common">
            <load>testbackground.png</load>
            <load>Edgeworth</load>
            <load>icon</load>
            <load>icon2</load>
            <load>discussion.mp3</load>
            <load>intense.wav</load>
            <load>talksoundmale.wav</load>
            <load>talksoundfemale.wav</load>
            <load>demonstration.conv</load>
        </group>
        <group name="intro">
            <load>intro0.0.png</load>
        </group>
    </groups>
```
Help me plzto get it

**Comment by wizered67** [comment]
I got an email notification but your comment isn't showing up so perhaps you figured out the problem and deleted it? Either way, the solution to the problem you were having is that the conversation needs to be loaded too! Add to the intro group `<load>intro.conv</load>`.

**Comment by OlegKostin** [comment]
Yes, ty) I figuried it out by myselft but thanks for worry anyway)
Hmm ive got another problem)
Can u center background picture plz?
cause now all images r starting to draw from left bottom corner and if u have portrait orientation pic it will know there, now at the center(

How do i use it     `<waitForLoading/>`?

```
     <loadGroup>atHome</loadGroup>
        <waitForLoading/>

```
Tried to use different ways( 

What for do i need        `<setScene name=""/>`?

Can i debug branches and convesations, by instant move to choosen one? This feature is in your another Hackathon game.
	public final static boolean DEBUG = false; im making it true and getting debug screen then push enter, gets that menu but still cant move to choosen branch or conversation. How should i do it? I  pushed confirm button after i had choosed the branch i wanted,  Still nothing had happened(
Im sorry that im bothering you, just wanna get how all it works.

Can i ` <unloadGroup>intro</unloadGroup>` in the same conversation? It crushing the game besause as i get it unload  <load>intro.conv</load> file too, though ive loaded it in another group before.

  ```
      <loadGroup>atHome</loadGroup>
        GameStory: ..........................

        <changebranch branch="atHome"/>
    </branch>
    <branch name="atHome">
        <bg id="intro1.0.png"/>

     Vika:   — ............................
        <unloadGroup>intro</unloadGroup>
```

```
       <group name="intro">
            <load>intro0.0.png</load>
            <load>intro0.1.png</load>
            <load>intro0.2.png</load>
            <load>intro0.3.png</load>
            <load>intro.conv</load>
            <load>heart.wav</load>
        </group>
        <group name="atHome">
            <load>intro.conv</load>
            <load>talksoundfemale.wav</load>
            <load>talksoundmale.wav</load>
            <load>heart.wav</load>
            <load>intro1.0.png</load>
            <load>intro1.1.png</load>

        </group>
```

**Comment by wizered67** [comment]
I've gone ahead and pushed a few changes that should hopefully fix any problems you were having.

> Can u center background picture plz?

I've added an optional x and y attribute to the background command so that you can set the position of the background. The position is still relative to the bottom left corner, but you can position it so that it appears centered to the player. The syntax by the way is `<bg id="image" x="0" y="0"/>` 

> How do i use it `<waitForLoading/>`?

The way you were using it is correct. Was that not working? I tested it and it works for me. The difference may be hard to notice if it loads resources quickly, but it just makes sure that any commands after it aren't processed until anything being loaded finishes loading.

> What for do i need `<setScene name=""/>`?

Scenes are optional and not necessary but could be helpful. Basically whenever you add character or images, you add them to the current scene. So you could add one set of characters to one scene, another set of characters to another, and then use setScene to switch from one scene to another without removing characters you don't need and adding the ones you do need every time.

> Can i debug branches and convesations, by instant move to choosen one?

I looked into it and tried to change the code a bit to make it easier. Once you switch to a branch, you will need to press enter to unpause the game and then possibly wait a few seconds, I'm not sure why yet. It should start the new branch after that.

> Can i `<unloadGroup>intro</unloadGroup>` in the same conversation?

There was a bug with unloading groups that caused a crash. I have fixed it now so try again. It should be fine to unload a conversation from within it.

Let me know if your problems are fixed or if you have any other problems!

**Comment by OlegKostin** [comment]
Everything is great, except debug mode.
`ERROR: Texture intro1.0.png was not yet loaded.`
It didnt work cause im loading this resourse later at code.
Thanks for your feedback.

**Comment by wizered67** [comment]
To be honest I'd say that's just a limitation of how debug mode works - it goes out of sequence when everything else assumes it has gone in sequence. You could add a branch that loads everything but is norally never switched to and then whenever you debug just switch to that first, unpause to let it load, and switch back to the branch you actually want. I could potentially make a debug way to load resources but that wouldn't solve other problems with branch switching, like that if you add a character earlier and then do something, the character will never be added if you switch to a later branch.

**Comment by OlegKostin** [comment]
Ok, save/load system seems is working. Thats nice.
Look ive got another problem) Cant change convesation.
intro.conv
```
        <loadGroup>preDayOne</loadGroup>
        <waitForLoading/>
        <changeconv conv="dayOne.conv"/>
    </branch>
</conversation>
```
dayOne.conv
```
<conversation xmlns="urn:conversations">
    <branch name="preDayOne">

        <bg id="intro0.0.png"/>
        <unloadGroup>atHome2</unloadGroup>
```
Recources.xml
```
 <group name="preDayOne">
            <load>intro0.0.png</load>
            <load>dayOne.conv</load>
            <load>heart.wav</load>
            <load>talksoundfemale.wav</load>
            <load>talksoundmale.wav</load>
        </group>
```

Exception in thread "LWJGL Application" java.lang.NullPointerException

I think this happened cause i didnt chose branch, i just dont know how.Help me plz

**Comment by wizered67** [comment]
Just a heads up that the save/load system isn't fully functional at the moment, there are still a couple things I need to fix with it. 

As for the issue, I really can't tell for sure based on what you said. Can you copy and paste the complete error message so I can see the line numbers?

**Comment by OlegKostin** [comment]
Here
```
Exception in thread "LWJGL Application" java.lang.NullPointerException
	at com.wizered67.game.conversations.ConversationController.setBranch(ConversationController.java:304)
	at com.wizered67.game.conversations.commands.ChangeConversationCommand.execute(ChangeConversationCommand.java:23)
	at com.wizered67.game.conversations.ConversationController.nextCommand(ConversationController.java:295)
	at com.wizered67.game.conversations.ConversationController.update(ConversationController.java:205)
	at com.wizered67.game.gui.GUIManager.update(GUIManager.java:165)
	at com.wizered67.game.screens.MainGameScreen.updateGUI(MainGameScreen.java:94)
	at com.wizered67.game.screens.MainGameScreen.render(MainGameScreen.java:83)
	at com.badlogic.gdx.Game.render(Game.java:46)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop(LwjglApplication.java:225)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:126)

```

**Comment by wizered67** [comment]
Okay, the problem is that when you change conversations it automatically starts at the "default" branch of the new conversation. You can fix the problem by adding a branch named "default" . However, I see why it could be useful to change to different branches of the new one so later today I'll add an optional attribute to change conversation to specify which branch to start on.

**Comment by OlegKostin** [comment]
ok ty

**Comment by wizered67** [comment]
Pushed the changes to the master branch! The syntax is `<changeconv conv="..." branch="..."/>`. Let me know if you need help with anything else!

**Comment by OlegKostin** [comment]
Can you add command, which will ask player to type anything and put the typing result to String variable, so you can use this variable in conversations, please? You said that u r working on that,i think this is important thing to add which will make life easier)

`   <if><![CDATA[ {work < 2} ]]></if>`
Which language is it?

> Variables
> 
> Variable tokens are replaced by any values you assign at runtime. e.g. you can create a text Hello {VAR=name}! and set the variable value with label.setVariable("name", "Mr. Bob").
> 
> Usage: {VAR=name}
> 
> name: Name of the variable that should be replaced at runtime.

This token thing u have added is already supporting variables. Nice.

**Comment by wizered67** [comment]
> Can you add command, which will ask player to type anything and put the typing result to String variable, so you can use this variable in conversations, please? You said that u r working on that,i think this is important thing to add which will make life easier)

I don't think this is something that needs it's own command. It's fully possible with the scripting system and I've added some example methods. Take a look at the getNameInput method in LuaGameMethods. It opens a text field and asks for the player's name, reopening if the player closes it. When calling it, you pass in the name of a Lua variable and it sets the result in that variable. Here's an example of how you could use it, as well as the setCharacterName method.

`<script>game:getNameInput("playerName")</script>`
`<waitUntil>playerName ~= nil</waitUntil>`
`<script>game:setCharacterName("Detective", playerName)</script>`

This will open a text input field for the player name and store the result in the variable named "playerName". It will wait until playerName is no longer nil (it gets set to nil at the beginning of getNameInput) and then set the character with the identifier "Detective"s name to the input. With some modifications I'm sure you can get the result you're looking for.

> <if><![CDATA[ {work < 2} ]]></if>
Which language is it?

The "work < 2" part is Lua. There's also the curly brackets { } around it because that's a requirement for if commands. I'm sure you're noticing the weird CDATA around it. In XML, anything inside a CDATA tag thing is read as plain text. This is necessary because otherwise the '<' sign is considered part of a tag which is a syntax error. Admittedly I'm not very happy with how ugly and confusing this is, so I'll consider  a cleaner solution.

> This token thing u have added is already supporting variables. Nice.

The new TypingLabel library I'm using adds variable tokens, but I actually still support the variable tokens I was originally using. You can embed a variable in a message with @v{name}.

**Comment by wizered67** [comment]
I've actually just changed the if command to be somewhat more intuitive. Now the syntax is `<if cond="lua statement">commands to execute here</if>`. Note that the condition code has been moved into an attribute named cond. This removes the need for having the curly braces around the condition which was annoying. Also, the '<' sign can now be used in conditions. The XML validator built into intelliJ will mark it as an error but running it in the game works fine!

**Comment by OlegKostin** [comment]
`        <waitUntil>playerName ~= nil</waitUntil>`
This part is bugged. You can cancel it sometimes. Sometimes not.

How can i use patterns for regular expression there? 
Cant understand how exactly getNameInput() works.

Thanks for your help!

**Comment by wizered67** [comment]
I haven't been able to reproduce this, unless you mean that hitting OK with no text inputted works? If that isn't what you're referring to, can you give more info on how to cancel it?

getNameInput just uses LibGDX's TextInput with a custom listener. The Listener just sets a Lua variable with the name passed into the original method equal to the text that the input method gets. Is your confusion with how the Lua variable is set or with how to use the TextInputListener? If it's with the listener take a look at https://github.com/libgdx/libgdx/wiki/Simple-text-input.

If you want to use regular expressions, you would use it on the text argument to the input method of the listener.

**Comment by wizered67** [comment]
Somewhat unrelated but I've added a new command which you may potentially find useful. The "while" command can be used to repeat a set of commands until a condition is no longer met, basically like a while loop in code. Here's an example.
```
<script>i = 0</script>
<while cond="i ~= 10">
     Detective: Hello! The iteration is @v{i}.
    <script>i = i + 1</script>
</while>
```

**Comment by OlegKostin** [comment]
> I haven't been able to reproduce this, unless you mean that hitting OK with no text inputted works? If that isn't what you're referring to, can you give more info on how to cancel it?

You can click on ok and it will continue even with empty textbox. If you click on cancel sometimes it canceled and u r getting empty background picture without anything else and ill get to restart app.
U can get thiss bug something if ull click on cancel once, something 2 or three times.
I think soemthing bad with logic but im not sure.

> Somewhat unrelated but I've added a new command which you may potentially find useful. 

Right now i cant see how i can use "while" in visual novel but i think someone can use it in future so god job.

What else are you going to implement in future? It seems to me that u can release your program,

**Comment by wizered67** [comment]
I've just made some changes that affect the scripting system and the name input, so please read!

In the interest of consistency, I've changed the syntax of `<waitUntil>` to use a "cond" attribute instead of putting the condition as text. The new syntax is `<waitUntil cond="condition script"/>`

Previously, the Lua variable "game" was mapped to the LuaGameMethods class, meaning you could add methods to LuaGameMethods and then call them with "game:methodName()". Now, the "game" variable is mapped to the GameManager class. The reason I did this is so that you can have access to more through Lua scripting. For example, you can access the GUI Manager with "game:guiManager()". If this doesn't seem important, don't worry about it. I've added a utils variable to GameManager which is assigned to LuaGameMethods, so you can still access the methods in the LuaGameMethods class with "game.utils:methodName()". 

Additionally, I've changed the getNameInput method in LuaGameMethods into a more generic getTextInput method. This allows you to use the same method for various use cases rather than just getting a name. getTextInput has one additional required argument where you can put the title of the popup window and also has two optional arguments to specify the default text to show and hint. Previously canceling the getNameInput method caused it to reopen. However, in some use cases it may be desired to let the user cancel an input. Therefore, I've changed it so that canceling sets the variable whose name was passed in to be equal to an empty string.

By using the new <while> command, it is still possible to require the player to input something! Here's the new way of getting the player's name.
```
<while cond=" playerName == nil or playerName == '' ">
    <script>game.utils:getTextInput("playerName", "Please enter your name")</script>
    <waitUntil cond="playerName ~= nil"/>
</while>
<script>game.utils:setCharacterName("Detective", playerName)</script>
```
This asks the player to input their name and waits until it is no longer nil, which will be the case as soon as the player enters something or hits cancel. If the player entered nothing or hit cancel, playerName will be equal to an empty string, so the while loop continues and the input is opened for the player again.

> If you click on cancel sometimes it canceled and u r getting empty background picture without anything else and ill get to restart app.

I haven't been able to reproduce this, both on desktop and my phone. Under certain circumstances on desktop the input field can end up behind the game window, but I can usually find it by minimizing or moving the game window. If there is a problem, it's most likely with LibGDX's text input. You could always add a TextField using Scene2D UI and get the input from that!

> What else are you going to implement in future? It seems to me that u can release your program,

I still have a todo list, but it's mostly bug fixes or minor features. I don't have any major features planned at the moment, unless I come up with more. I may go back to the resource system and see if I can improve it. It's functional, but your confusion with it proves that it isn't perfect. I also want to try to make the command syntax consistent and logical and add documentation for them at some point. Other than that, please give me your feedback if there are features you think should be added or bugs you find. I'll be looking forward to seeing what you come up with!

**Comment by OlegKostin** [comment]
Everything works fine,I havent found any bugs yet.

`or playerName == '' ">`
is this  regular expression? If not can i add it there?
And i like your getTextInput() method in LuaGameMethods, ive made something like that by myself but yours is really  better.

Are you going to add nice gui for load/save system? Cause i dont know how to add.

**Comment by wizered67** [comment]
> is this regular expression? If not can i add it there?

It's not really a regular expression, it just checks if the string is empty. Lua does have regular expressions you could add there, but personally I think it would be easier to use Java regular expressions. You could do this by adding a method to LuaGameMethods that takes a String and returns whether it matches the regular expression you want. Then, just add a call to that to the while command condition, ex:
`<while cond=" playerName == nil or playerName == '' or ~game.utils:isValidName(playerName)">` where isValidName is the name of the method you add to LuaGameMethods.`

> And i like your getTextInput() method in LuaGameMethods, ive made something like that by myself but yours is really better.

Thanks, I'm glad you like it :D

> Are you going to add nice gui for load/save system? Cause i dont know how to add.

I may add one in the future but it isn't really a priority. Some stuff is better left to individual game creators to suit their game best. Are you familiar with making UIs with Scene2D UI? You can pretty much just make your UI and add some code to GUIManager for it to work.

**Comment by wizered67** [comment]
I've added a couple small features and made some minor changes the last couple days that I wanted to update you on in case you want to use them.

**Screenshake**
I've added a new command for applying a screenshake effect. The syntax is `<screenshake time="" xMagnitude="" yMagnitude="" interpolation="optional" wait="optional"/>` Time changes how long the screenshake lasts and xMagnitude and yMagnitude are up to how many pixels in either direction the screen can shake. If wait is true, the screenshake must finish before the next commands are processed. Interpolation is an optional attribute, linear by default, that affects the dampening of the screenshake. 

**Music Indices**
Previously, playing music would stop any music already playing. However, there may be some niche cases where it is desirable to have multiple music playing at once, for example if some is used for ambient noise. Therefore, all music related commands (except playsound, which uses sounds) have an extra optional index attribute, 0 by default. Here's an example of how it works. If you play music on index 0 and then play another track on index 0, the first one stops. But if you play the second one on index 1, whatever was playing on index 1 stops and is replaced by the new one, but index 0 continues too. This works similarly with pausing and resuming, where only the track at the index is affected.

**Scale Command**
There's now a scale command which can set the scale of a sprite for a character or image. This command replaces the previous direction command which was a more limited version, only letting x be flipped.  The syntax for scaling a character is `<scale id="" xScale="" yScale="">` and the syntax for scaling an image is adding into an `<image>` block a `<scale xScale="" yScale="">`.

I hope these changes are helpful, either now or in the future. There were also some small bug fixes done. Let me know if I broke anything or if you have any questions!

**Comment by OlegKostin** [comment]
I think i need to repeat some parts of libgdx ,but by the way. How can i add background image over         ```
Gdx.gl.glClearColor(0x64 / 255.0f, 0x95 / 255.0f, 0xed /
                255.0f, 0xff / 255.0f);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
```
but before actual background from conversation. So if some conversations background r smaller then screen, player will see part of another picture not a red background.
Screenshake is nice,
 what is syntax for new music tags?

**Comment by wizered67** [comment]
It sounds like you just want to add another background below the main one? That's already possible using images. I thought about allowing multiple backgrounds like multiple music but I'm not sure the point. As it is, the background is actually just a slightly shorter way of adding an image, so maybe I should just remove it and have people use the image tag? 

For music tags, you can just add an index attribute to any of them, ie `<resumemusic index=1/>`

**Comment by OlegKostin** [comment]
I think u should keep background tag.  And you are right, i did it wrong, i can make multiple backgrounds already.

**Comment by wizered67** [comment]
Yeah, I guess it makes sense to keep the background tag as a simple shortcut for the most common use case of backgrounds. If people need something more complicated they can just use images. My only concern is that having a background tag would make people think backgrounds are completely separate from images when really backgrounds are just images.

I've been putting some thought into how to improve the resource system to be simpler. I would love your thoughts if you have some. I'm currently thinking something along the lines of the ResourceSchemaGenerator.jar automatically looking through all files in the assets directory so that you don't have to list out every single one in Resources.xml, although you'd still have to include them in load groups. I'm really trying to make this with developers in mind so please tell me what the ideal resource system is like for you!

**Comment by OlegKostin** [comment]
At first i think u need to split  Resources.xml into 2 files, one for loading groups and another for initialization of them.
And at second you need to add image atlases like u did for animation. But still support non atlases images.
At third u need to add directory at assets with name "init" or "initialization " and split resource  initialization into files like    "textures " "sounds" "animations" and so on.
 I think this idea will help developers to avoid mistakes and make it easy to add new filesю
Because right now my Resources.xml has 537 lines and i think this is so wrong.

**Comment by wizered67** [comment]
Thank you for your feedback! It will definitely be taken into consideration.

**Comment by OlegKostin** [comment]
Ive found the bug.
Debug doesnt work with conversations. Ive loaded all groups at start, but i can choose one first conversation branches. If ill choose another conversation it will be crushed.

**Comment by wizered67** [comment]
I believe I should have fixed the issue. Keep in mind you'll have to set the branch you want to use after setting the conversation.

**Comment by wizered67** [comment]
I've been thinking about the resource system and have come up with the following scheme. Before I start implementing it, I wanted to get your opinion on it.

The assets folder will contain 3 files - a resources list, a loading list, and a character list - as well as a number of folders. The loading list file will essentially be the bottom part of the current resources.xml. It will contain all the groups and you list within the groups which assets to load. The character list will also look very similar to the character definitions part of resources.xml. The resources list will contain the names of all subdirectories in the assets folder and the class name of the assets they contain. With this system, the user can add as many folders as they want for different assets because they can specify the types themselves. This is different from before where it was hardcoded that all assets in the "Music" directory were the Music type. 

Here's an example of what the resources list might look like:
```
<directory name="Music" type="com.badlogic.gdx.audio.Music"/>
<directory name="GUI Textures" type="com.badlogic.gdx.graphics.Texture"/>
<directory name="Backgrounds" type="com.badlogic.gdx.graphics.Texture"/>
.......
```

The idea is to let users structure their directories in whatever way is most convenient for them, without having to follow some format I impose. Note how it allows the textures to be split into two folders which previously would have required code changes for every additional folder. This also lets users add resources of custom types since they get to specify the type of the directory.

Inside each of these subfolders there will be a new resources.xml file. The file will contain xml tags for each resource in the folder and will have an identifier attribute for specifying a custom identifier. If not specified, it will be the same as the file name. Each resource element can also have some nested elements to give it properties. These will be hardcoded for specific types. For example, animations will have an element to give it properties such as animation speed. Music will have an element to give it a custom volume to be played at. 

One change is that instead of manually typing the names of all resources you want to add to resources.xml, running the automatic generator program will add all files within the directory to resources.xml in order to save time.

Here's an example of what resources.xml could look like in a music directory:
```
<music filename="questioningMusic.mp3" identifier="questioningMusic.mp3">
  <volume>0.5</volume>
</music>
<music filename="intensemusicFreeSrc(5).mp3" identifier="intenseMusic"/>
```
Part of the idea is that common resource types like music will have these custom elements like "music" so that they can contain the necessary properties - like volume.

Another change that will be made is allowing textures to be stored in atlases. At this point it will probably look something like how animations work, so accessing the image "candle" in atlas with identifier "Objects" may look like "Objects_candle". Just using regular textures will also be supported. As part of this change I'm planning on allowing images to have animations as well, including backgrounds of course. 

I think that basically summarizes my design plans. I think one of the most important changes is letting users structure their directories however they want, so that if one becomes large they can split it into several. This whole system may take a while to implement because it is somewhat intricate and I want to make it as easy to use as possible. Please let me know your thoughts!

**Comment by OlegKostin** [comment]
sounds really nice. And i really like the automatic generator program idea.)

**Comment by OlegKostin** [comment]
sorry ive figured it out by myselft.

Can u explain anyways when do we use 
     `<if cond="var"/>`?
Is it useful only in choices?
Cause im using  
```
<if cond="var">
code
</if>
```

**Comment by wizered67** [comment]
Oh, glad you figured that out. Yeah, `<if cond="var"/>` is only used as the first line in a choice and it makes it so that the choice only appears if that condition is met. I agree it's confusing but I'm not sure right now what a better solution would be. Perhaps I should switch back to using xml tags for choices like it originally was. For example, instead of the current 
```
<choices>
  Choice One:
    Text
  Choice Two:
    Text
</choices>
```
 it used to look like
```
<choices>
  <choice option="Choice One">
    Text
  </choice>
  <choice option="Choice Two">
    Text
  </choice>
</choices>
```

If I went back to this way I could just add a condition attribute to each `<choice>` tag. The reason I originally switched to this new way was that I thought it looked a bit too cluttered with all the extra `<choice>` tags. Do you have any preferences?

**Comment by OlegKostin** [comment]
Im not sure but its not quite obviously what "if" condition runs after choice text. I think your tags way is more obvious.

**Comment by OlegKostin** [comment]
Hmm ive got another problem.
Ive made my own save load and exit buttons in guimanager, it wasnt hard. 
But why cant i add listeners for buttons in another screens?
Do i need to register them in guimanager somehow?
As i understand buttons in guimanager r working for MainGameScreen.
Maybe im doing something wrong, but i cant add any listeners for buttons in another screen, plz help me)

**Comment by OlegKostin** [comment]
This code is long, sorry, i tried that way but it doesnt work(

```
public class MenuScreen implements Screen, Controllable {
    private Stage stage;
    private Image imgBackground;
    Texture bg;
    Texture startButton;
    TextureRegion stBtn;
    TextureRegionDrawable startB;
    private Button btnMenuPlay;
    MainGameScreen gameScreen;

    @Override
    public void show() {

        stage = new Stage(new
                StretchViewport(Constants.VIRTUAL_WIDTH,
                Constants.VIRTUAL_HEIGHT));
        rebuildStage();
    }

    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0x64 / 255.0f, 0.0f, 0.0f, 1.0f);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
        rebuildStage();
        stage.act(delta);
        stage.draw();
    }

    private Table buildControlsLayer() {
        Table layer = new Table();
        layer.right().bottom();
        btnMenuPlay = new Button(startB);
        layer.add(btnMenuPlay);
        btnMenuPlay.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                onPlayClicked();
            }
        });
        layer.row();

        return layer;
    }

    private void onPlayClicked() {
        gameScreen = new MainGameScreen();
        GameManager.game.setScreen(new LoadingScreen(new LoadingScreen.LoadResult() {
            @Override
            public void finishLoading() {
                GameManager.game.setScreen(gameScreen);
                GameManager.conversationController().setConv(GameManager.assetManager().getConversation("intro.conv"));
                GameManager.conversationController().setBranch("intro");
            }
        }));

    }

    private Table buildBackgroundLayer() {
        Table layer = new Table();
        // + Background
        imgBackground = new Image(bg);
        layer.add(imgBackground);
        return layer;
    }

    private void rebuildStage() {
        bg = new Texture(Gdx.files.internal("Textures/bg.png"));
        startButton = new Texture(Gdx.files.internal("start.png"));
        stBtn = new TextureRegion(startButton);
        startB = new TextureRegionDrawable(stBtn);
        Table layerBackground = buildBackgroundLayer();
//        Table layerObjects = buildObjectsLayer();
//        Table layerLogos = buildLogosLayer();
        Table layerControls = buildControlsLayer();
//        Table layerOptionsWindow = buildOptionsWindowLayer();
// assemble stage for menu screen
        stage.clear();
        Stack stack = new Stack();

        stack.setSize(Constants.VIRTUAL_WIDTH,
                Constants.VIRTUAL_HEIGHT);
        stack.add(layerBackground);
//        stack.add(layerObjects);
//        stack.add(layerLogos);
        stack.add(layerControls);
        stage.addActor(stack);
//        stage.addActor(layerOptionsWindow);
    }

    @Override
    public void resize(int width, int height) {

    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {
        stage.dispose();
    }

    @Override
    public void dispose() {
        stage.dispose();
    }

    @Override
    public void touchEvent(int screenX, int screenY, int pointer, int button, boolean pressed) {

    }

    @Override
    public void keyEvent(MyInputProcessor.ControlType control, int key, boolean pressed) {

    }
}

```

**Comment by wizered67** [comment]
Okay, I figured out the problem, although I'm not entirely sure the best way to fix it yet. As I've been developing this I kind of just focused on having the main screen and didn't put much thought into how to handle having other screens, which admittedly is a big mistake and something I'll try to fix in a future version.

For now, I can give you a solution that should work but isn't the best. Go to GameManager and add the following method:
```
public static void setDefaultInput() {
  Gdx.input.setInputProcessor(game.inputMultiplexer);
}
```

In your Screen's show method, after creating your stage add `Gdx.input.setInputProcessor(stage);` and in the "finishLoading" part you create in onPlayClicked, add a call to `GameManager.setDefaultInput()`

Again, sorry this is much more complicated than it should be. I made some bad assumptions while working on the project so that I could prioritize working on conversations. I'll definitely try to fix this in a future version!

**Comment by OlegKostin** [comment]
`   GameManager.setDefaultInput();`
Where should i place it? Didnt get it, sorry) But u see im useful im finding bugs)

**Comment by wizered67** [comment]
Here, change your onPlayClicked to this.

```
private void onPlayClicked() {
        gameScreen = new MainGameScreen();
        GameManager.game.setScreen(new LoadingScreen(new LoadingScreen.LoadResult() {
            @Override
            public void finishLoading() {
                GameManager.setDefaultInput();
                GameManager.game.setScreen(gameScreen);
                GameManager.conversationController().setConv(GameManager.assetManager().getConversation("intro.conv"));
                GameManager.conversationController().setBranch("intro");
            }
        }));

    }
```

**Comment by OlegKostin** [comment]
Yes i tried that way but it doesnt work for me(

**Comment by wizered67** [comment]
Hmm... not sure why it isn't working. I'll test it and figure it out!

**Comment by OlegKostin** [comment]
Ok, maybe ive made miskates somewhere at this screen. But i think the main problem is ChangeListener. Ive be glad if ull help me.

**Comment by wizered67** [comment]
Ah, I found the problem. Right now in the render method you keep calling `rebuildStage();`
That shouldn't be necessary since you only need to create the stage once. Calling it every frame seems to be messing with the listener. See if that fixes the problem!

**Comment by OlegKostin** [comment]
Oh thanks!! It was stupid mistake!  Now its working, thanks

**Comment by OlegKostin** [comment]
Can u give me plz an advice? Ive tried my game on real android device, Its going fine but after loading and unloading six groups its crushing cause of low memory.
Is it possible that it doesnt dispose any recources somehow? Everything is fine with desktop  but here is a problem with android.(

**Comment by wizered67** [comment]
That's unfortunate. I haven't tested it before but I assumed it wouldn't be an issue because it uses an AssetManager. I will take a look at this and try to figure out the problem in the next few days! Can you give me some idea about how large the resources you load are for each group? Also, on desktop is the memory increasing but just not reaching the limit?

**Comment by OlegKostin** [comment]
> Also, on desktop is the memory increasing but just not reaching the limit?

Hmm no, memory isnt increasing that much, just on 0,5 percent, i think i just need to split my groups on even smaller pieces. Thanks)
Waiting for your ugrades! )

**Comment by wizered67** [comment]
Okay, well I'll look into it! I've been working on the improved resource system and it's nearly usable! It definitely should be an improvement. I'll let you know when it's done and give some instructions on how to use it!

**Comment by wizered67** [comment]
I've completed a working version of the new resource management system and pushed it to the master branch! In the assets directory, you'll see 3 XML files and a new and improved ResourceSchemaGenerator.jar. 

The first XML file is CharacterDefinitions.xml. This one should be pretty simple - it's essentially the characters section you're used to using. You can list characters using the `<character>` tag and specify their id, default name, and speaking sound effect.

The second XML file is LoadingGroups.xml. Again, it's very similar to the load groups section you're used to using. You specify the names of groups and then list the resources to load/unload in each group.

The most important of the XML files is ResourceDirectories.xml. This one lists the directories that contain resources. In each `<directory>` tag, you put the name of the directory and it's type, which is essentially the full name of the Java class it corresponds to, including packages. I think figuring out the type to put should be pretty straight forward with one possible exception - if a directory contains TextureAtlases, whether they contain animations or static images, you should use the type `com.wizered67.game.assets.AnimationTextureAtlas`. (Another quick side note, each directory tag can have multiple exclude tags, which means resources with certain extensions should not be processed. For example, with atlases you want to process the .pack files but not the .png files, so add .png to exclude tags).

Speaking of, as a brief side note, images can now be stored in TextureAtlases. In fact, I've essentially removed the line between animations and images. Character animations and image graphics can now both either be texture files, animations from a texture atlas, or static images from a texture atlas. Referencing textures from an atlas is done in the same syntax as animations - AtlasIdentifier_RegionName. As part of this change I've modified some syntax. Setting the graphic of an Image can now be done with 
```
<image>
  <graphic id="texture/texture from atlas/animation from atlas"/>
</image>
```
The graphic tag was previously called "texture". I was going to also change the `<animation>` tag but decided not to so that you don't have to change your conversation files too much.

Okay, let me give some info on how to fully utilize the new resource management system. First, you may need to go to your IDE's "Schemas and DTDs" window and add the new ResourceConfigSchema.xsd. I removed one of the old schemas so you may need to remove it to.

Using the resource system should be pretty simple. Simply add the directories to ResourceDirectories.xml and then add your resources as normal to each directory. Then, run ResourceSchemaGenerator.jar. You may need to set its working directory to the assets folder. Running ResourceSchemaGenerator.jar will automatically make a Resources.xml file in each directory. It automatically contains a list of all resources in the directory with their names and identifiers. You can optionally change the identifiers to something different for convenience, as you could before. Some types of resources will have special configuration info to provide. For example, textures have a special `<texture>` element to specify things like filters to use. For animations, you can provide info on speed and play mode. After modifying these configurations, run ResourceSchemaGenerator.jar again. Also be sure to run it after modifying CharacterDefinitions.xml or LoadingGroups.xml.

Hopefully this new system will make things much easier since you can split things into different directories however you want, as long as each directory only has one type for all its files. Also, a lot of things are now automatically generated to save you some work! Since this is a new system, there may be bugs, and in the worst case a Resources.xml file could be lost if an error occurs, although I've tried to prevent this. To prevent you from losing anything important, each time ResourceSchemaGenerator.jar runs it will generate a backup of every directory's Resources.xml file. If things go wrong, check the backup folder before continuing!

I'd like to think things are pretty self explanatory but some of it might be confusing. Let me know if you have questions! Overall, I think this will be a very positive change. I apologize if my syntax tweaking has caused you to need to make changes to your files. Good luck and enjoy the changes!

**Comment by OlegKostin** [comment]
Ok good job!! I ll try it out. And by the way
`   GameManager.setDefaultInput();`  
What does it mean? Do we have two input prossers? 
I removed that code and nothing have changed) So its very strange) 

And also ive finished my test game and now i wanna make nice kid game based on story was written by my wife. It will be on my native language. Can i use your engine? It will be free but i hope ill publish it at playmarket. Also i want to add you to credits.
 It will be not that big but with our own pictures and text. Just small  fairy tale.

**Comment by wizered67** [comment]
Hmm... I could have been mistaken about needing setDefaultInput. There is a separate input processor for the main game so I thought it would be necessary to switch back to that when the screen changes.

I don't have an official license for this yet, but you have my permission to use the engine for your project. Publishing it is fine, just let me know if you plan on making it paid. Thank you for asking!

**Comment by wizered67** [comment]
I did a little testing to look into the out of memory problem you mentioned. I tried repeatedly loading a group, doing a little dialogue, and then unloading the group, repeating this process over and over again. As far as I could tell the memory usage on my Android phone was not increasing, so I believe there isn't an issue with resources not being disposed. My guess is that you had the problem because you had too much loaded at once, but I'm not certain. If the issue keeps happening let me know and I can look into it further!

**Comment by OlegKostin** [comment]
Can u give me an advice plz?
 Desktop version is working just fine, but android one is crushing if im starting not from LoadingScreen or MainGameScreen. 
```
Warning:[options] bootstrap class path not set in conjunction with -source 1.7
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(luajc$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$10) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$2) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$3) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$4) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$5) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$6) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$7) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$8) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.CoerceJavaToLua$9) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.JavaArray$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.JseProcess$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.lib.jse.LuajavaLib$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.luajc.BasicBlock$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.luajc.BasicBlock$2) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.luajc.BasicBlock$3) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.luajc.VarInfo$1) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Error:warning: Ignoring InnerClasses attribute for an anonymous inner class
Error:(org.luaj.vm2.luajc.VarInfo$2) that doesn't come with an
Error:associated EnclosingMethod attribute. This class was probably produced by a
Error:compiler that did not target the modern .class file format. The recommended
Error:solution is to recompile the class from source, using an up-to-date compiler
Error:and without specifying any "-target" type options. The consequence of ignoring
Error:this warning is that reflective operations on this class will incorrectly
Error:indicate that it is *not* an inner class.
Information:BUILD SUCCESSFUL
Information:Total time: 1 mins 23.942 secs
Information:152 errors
Information:1 warning
```
Also ive got many errors in gradle. Why can it be? Can it be cause of inputProcessor? 
As i said desktop version is sworking just fine even from MenyScreen and all buttons r working there fine.

**Comment by wizered67** [comment]
Can you give more info on what the crash is when you run it on Android, preferably the full error message. The gradle errors are "normal". They say they're errors but they're actually just warnings, so you can ignore them. It has to do with the Lua library, apparently it was compiled on an old version of Java or something. They shouldn't cause problems so just ignore them.

Also I did some fixes for the InputProcessor stuff. It's actually a minor change to the original code but should make things a little easier. I can push the code soon but you'll have to make a few minor changes to adjust to it.

**Comment by OlegKostin** [comment]
I figured it out. I had updated another stage)
So gogo waiting for your ugrades, you r going perfect job!
I guess this is my problem, android cant read json file with skin for some reason, Its working on desktop though.
```
rocess: com.wizered67.game, PID: 2300
                                                                  com.badlogic.gdx.utils.SerializationException: Error reading file: images/uiskin.json
                                                                      at com.badlogic.gdx.scenes.scene2d.ui.Skin.load(Skin.java:98)
                                                                      at com.badlogic.gdx.scenes.scene2d.ui.Skin.<init>(Skin.java:83)
                                                                      at com.wizered67.game.screens.MenuScreen.rebuildStage(MenuScreen.java:328)
```
Ok, i solved that problem too, It was a problem with font. Now everything is working! Nice. Im starting project for kids now. Hope it will be good one. Uve made a really nice engine!

**Comment by OlegKostin** [comment]
Look, ive found a bug.
getTextInput doesnt work with Fullscreen mode.
```
   <while cond=" playerName == nil or playerName == '' ">
            <script>game.utils:getTextInput("playerName", "text","","text")</script>
            <waitUntil cond="playerName ~= nil"/>
        </while>
```

This command will alt+tab the whole game and make u type the name on  your Desktop. Can u fix it plz?

**Comment by wizered67** [comment]
I'm glad you figured out the other problems you were having! Maybe later today/tomorrow I can push the changes I mentioned with the new input processor stuff. I also contacted the creator of the library being used for the scrolling text label used for displaying dialogue and he's working on fixing the last couple problems I was having with it, so hopefully those changes will be ready soon.

As for your getTextInput issue, I'm pretty sure that's just how the default LibGDX text input works on desktop. If you want to fix that, I'd recommend making a UI with Scene2D for text input, probably using Scene2D TextField. I can maybe take a shot at it later if you want, maybe next week.

I'm glad you like the engine! Have you tried out the new resource management system yet and do you like it?

**Comment by OlegKostin** [comment]
Sorry i havent tried it yet, i was working on gui of my test project, but tomorrow ill start working on my new prject which will base on your new resource management system.
 And i hope ull fix my problem with getTextInput. But if not ill try to fix it later.

**Comment by wizered67** [comment]
I've gone ahead and merged into the master branch the changes I've been working on to the input system. You can see MainGameScreen to see how I use it, but here's the general idea. 

There is one global InputMultiplexer that is always set as the active input processor by LibGDX. You can add your InputProcessors/InputMultiplexers to this one with GameManager.addInputProcessor(InputProcessor, (optional) index) and passing in the processor and optionally the index in the global multiplexer to set it to. When you're done with it, you can remove it with GameManager.removeInputProcessor(InputProcessor). What I did in the new MainGameScreen and what I'd recommend doing is initializing that screen's input processors in its constructor, adding them in the screen's show method, and removing them in the hide method. 

I've added a special InputProcessor called ControlInputAdapter. You can create a ControlInputAdapter and pass in a class implementing the Controllable interface. The Controllable interface is supposed to be a minimalist version of InputProcessor that only requires implementing touchDown, touchUp, keyDown, and keyUp. The difference between these methods and the regular ones, and also the purpose of ControlInputAdapter, is that the keyDown methods take an extra argument corresponding to a ControlType. The idea is that for keyboard inputs, it makes sense to allow multiple keys to map to a ControlType. For example, allowing both "W" and the "Up" key for an Up ControlType. This could also allow in the future for letting players customize key bindings.

I may make some changes to this input system in the future, but this seems to work for now. Look at MainGameScreen if you want to see an example! And let me know if you have questions or find bugs.

**Comment by wizered67** [comment]
Today's your lucky day! I just added a new text input UI. It's nothing special and isn't particularly pretty, but it works. Adding it reminded me of some problems with the UI system so I might make some changes in the future. Using it should be about the same except that I removed one optional argument from game.utils:getTextInput. I couldn't figure out how to use the hint argument so it has been removed. Try out the new UI and let me know what you think!

**Comment by OlegKostin** [comment]
```
<while cond=" playerName == nil or playerName == '' ">
    <script>game.utils:getTextInput("playerName", "Please enter your name")</script>
    <waitUntil cond="playerName ~= nil"/>
</while>
```
look, u can just click somewehre not at this window and condition will be ok cause  playerName == '' " and not a null. So this one doesnt work. And whole gmae will be failen(

**Comment by wizered67** [comment]
Hi OlegKostin, can you clarify the problem you're having? I tested it and it seems to work as intended. 

There is one potential problem in that if the player enters one or more spaces it will accept it even though the name is empty. I've pushed some changes that fix that problem. Take a look at the updated example in demonstration.conv. You can also modify the isValidTextInput method in LuaGameMethods to further restrict input, including using regular expressions to do so!

Let me know if there is still a problem and the specifics on how to reproduce it!

**Comment by OlegKostin** [comment]
i didnt update your last commit. U have fixed it already, so im sorry that i bothered you.
But how can i add hint there? As i get u dont know how.(
How can i modify TextInputUI? Add size and stuff.  Just didnt get how can i work with that to make it looks better(

**Comment by wizered67** [comment]
Just figured out that LibGDX's text field already supports hints, so I added them back. Make sure to pull the newest commit. TextInputUI is just composed of several Scene2DUI labels/buttons created in the constructor. You can modify it like any other UI made with Scene2D.

**Comment by OlegKostin** [comment]
Hello again! I need some your help..
 Look is it possible to make that ill have some button and by clicking on it ill get random branch from choosen conversation? I want to add also variables which will deny some branches until player will unlock them. Thanks!!

**Comment by wizered67** [comment]
Hello! First, I wanted to let you know that I won't be doing any development on this project for a couple months. I will resume around the end of August. 

As for your question, there should definitely be a way to do what you're looking for. I won't write the code but I can point you to the methods that you'll want. First, you'll need a reference to the Conversation object with the branches you're trying to use. If it's the current conversation, you should be able to get it with `GameManager.conversationController().conversation()`. If it's not the current conversation, you can load it and then get it through the asset manager with `GameManager.assetManager().get("conversation identifier", Conversation.class)`. You probably should also switch to that conversation if it isn't the current one using `GameManager.conversationController().loadConversation(conversation)`.

Once you have the Conversation object, you can get a set with the name of all of it's branches with `conversation.getAllBranches()`. I would probably then iterate over all the branch names and for each one check if it should be a potential branch to switch to. If it is, add it to a list, and once you've done this for all branches just choose a random one from the list. Then, you can set the branch with `GameManager.conversationController().setBranch("branch name")`. You may need to manually call `GameManager.conversationController().nextCommand()`, I don't remember.

Okay, the last thing you mentioned is that you want to choose which branches to deny based on variables. If you mean Java variables that should be very simple - just check them when iterating over all branch names and don't add to the list if they aren't what you want. If you mean Lua variables you can mostly do the same thing, but you'll need to get the Lua variable. You can get a Lua variable with `GameManager.conversationController().scriptManager("Lua").get{Integer/Double/String/Boolean}Value("variable name")`.

Hope this helps, let me know if you have any questions.

**Comment by OlegKostin** [comment]
Ok thanks! And please continue your project at the end of August, Its really great. And i think ill will ask you more question. Thanks for support.

**Comment by wizered67** [comment]
Hello! I'm planning on resuming development in the next week or so. I believe my priority for now will be making this more of a library that can be used in a variety of projects. Right now to use this you kind of have to build your project around it. My vision is to make this a dialogue system that can be added to many types of games, but would be especially good for visual novels. For now I will probably be more focused on cleaning up code and making it more accessible than adding new features. However, if there are any features you would like added, let me know and I'll add them to my list for the future. Also, I was wondering if you've been using the new resource management system and whether it has been working well for you?

**Comment by OlegKostin** [comment]
Hello, yes its working. Hope ull continue your project. Its nice one.
